src/main/java/com/faforever/api/map/MapService.java

```diff
@@ -0,0 +1,348 @@
+      ".scmap",
+      "_save.lua",
+      "_scenario.lua",
+      "_script.lua"};
+  public static final Charset MAP_CHARSET = StandardCharsets.ISO_8859_1;
+  private final FafApiProperties fafApiProperties;
+  private final MapRepository mapRepository;
+  private final ContentService contentService;
+
+  @Inject
+  public MapService(FafApiProperties fafApiProperties, MapRepository mapRepository, ContentService contentService) {
+    this.fafApiProperties = fafApiProperties;
+    this.mapRepository = mapRepository;
+    this.contentService = contentService;
+  }
+
+  @Transactional
+  public void uploadMap(byte[] mapData, String mapFilename, Player author, boolean isRanked) throws IOException {
+    if (author == null) {
+      throw new IllegalStateException("author cannot be null");
+    }
+    if (mapData.length <= 0) {
+      throw new IllegalStateException("mapData is empty");
+    }
+    MapUploadData progressData = new MapUploadData()
+        .setBaseDir(contentService.createTempDir())
+        .setUploadFileName(mapFilename)
+        .setAuthorEntity(author)
+        .setRanked(isRanked);
+
+    progressData.setUploadedFile(Paths.get(progressData.getBaseDir().toString(), mapFilename));
+    copyToTemporaryDirectory(mapData, progressData);
+
+    unzipFile(progressData);
+    postProcessZipFiles(progressData);
+
+    parseScenarioLua(progressData);
+    checkLua(progressData);
+    postProcessLuaFile(progressData);
+
+    updateMapEntities(progressData);
+
+    renameFolderNameAndAdaptLuaFiles(progressData);
+    generatePreview(progressData);
+
+    zipMapData(progressData);
+    cleanup(progressData);
+  }
+
+  private Path copyToTemporaryDirectory(byte[] mapData, MapUploadData progressData) throws IOException {
+    return Files.write(progressData.getUploadedFile(), mapData);
+  }
+
+  @SneakyThrows
+  private void unzipFile(MapUploadData mapData) {
+    try (ZipInputStream zipInputStream = new ZipInputStream(new BufferedInputStream(
+        Files.newInputStream(mapData.getUploadedFile())))) {
+      Unzipper.from(zipInputStream).to(mapData.getBaseDir()).unzip();
+    }
+  }
+
+  @SneakyThrows
+  private void postProcessZipFiles(MapUploadData mapUploadData) {
+    Optional<Path> mapFolder;
+    try (Stream<Path> mapFolderStream = Files.list(mapUploadData.getBaseDir())) {
+      mapFolder = mapFolderStream
+          .filter(path -> Files.isDirectory(path))
+          .findFirst();
+    }
+
+    if (!mapFolder.isPresent()) {
+      throw new ApiException(new Error(ErrorCode.MAP_MISSING_MAP_FOLDER_INSIDE_ZIP));
+    }
+    mapUploadData.setOriginalMapFolder(mapFolder.get());
+    mapUploadData.setUploadFolderName(mapUploadData.getOriginalMapFolder().getFileName().toString());
+
+    List<Path> filePaths = new ArrayList<>();
+    try (Stream<Path> mapFileStream = Files.list(mapUploadData.getOriginalMapFolder())) {
+      mapFileStream.forEach(myPath -> filePaths.add(myPath));
```

Comment:
Use method references when IntelliJ suggests it. But, the code can be reduce to::

```
    try (Stream<Path> mapFileStream = Files.list(mapUploadData.getOriginalMapFolder())) {
      List<Path> filePaths = mapFileStream.collect(Collectors.toList());
      Arrays.stream(REQUIRED_FILES)
          .forEach(filePattern -> {
            if (filePaths.stream()
                .noneMatch(filePath -> filePath.toString().endsWith(filePattern))) {
              throw new ApiException(new Error(ErrorCode.MAP_FILE_INSIDE_ZIP_MISSING, filePattern));
            }
          });
    }
```